[WIP] Support lnd format in ChannelGraph

VasanthManiVasi opened the pull request:
This adds support for the ChannelGraph in lnd format.

It converts the lnd format to cln format and then loads the graph

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -13,22 +13,29 @@ class ChannelGraph():
     """
 
-    def _get_channel_json(self, filename: str):
+    def _get_channel_json(self, filename: str, fmt: str = "cln"):
```
I like how you made it backwards compatible!

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -13,22 +13,29 @@ class ChannelGraph():
         """
-        f = open(filename)
-        return json.load(f)["channels"]
+        with open(filename) as f:
```
just curiosity is there any rational why the `with ... as ...` syntax is preferable?

renepickhardt replied:
Will try to get used doing this. this is an excellent reason!

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -46,3 +53,66 @@ def get_channel(self, src: str, dest: str, short_channel_id: str):
+            "short_channel_id": lnd2cln_channel_id(int(lnd_channel["channel_id"])),
+            "satoshis": int(lnd_channel["capacity"]),
+            "amount_msat": int(lnd_channel["capacity"]) * 1000,
+            "public": None, # Info not available in lnd describegraph?
```
Not entirely sure. I guess it is public if it is in the json? guess we have to check the lnd code repo or ask someone there

VasanthManiVasi replied:
Thanks, that makes sense 👍
It's set to True in a174641

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -46,3 +53,66 @@ def get_channel(self, src: str, dest: str, short_channel_id: str):
+    Convert from lnd short channel id to cln short channel id
+    """
+
+    # https://bitcoin.stackexchange.com/a/79427
```
nice quoting. wonder if this should be moved to the doc string?

VasanthManiVasi replied:
Done in a174641

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -46,3 +53,66 @@ def get_channel(self, src: str, dest: str, short_channel_id: str):
+
+def lnd2cln_channel_id(lnd_channel_id: int):
+    """
+    Convert from lnd short channel id to cln short channel id
```
actually cln uses the `short_channel_id` format / encoding as defined in the BOLTs. I am not sure why lnd never switched.

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -46,3 +53,66 @@ def get_channel(self, src: str, dest: str, short_channel_id: str):
+
+    return cln_channel_json
+
+def lnd2cln_channel_id(lnd_channel_id: int):
```
see next comment. Maybe `lnd2bolt_short_channel_id`?

VasanthManiVasi replied:
Done in a174641

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -46,3 +53,66 @@ def get_channel(self, src: str, dest: str, short_channel_id: str):
+    }
+
+    def _add_direction(src, dest, lnd_channel, cln_channel):
+        cln_channel["source"] = lnd_channel[src + "_pub"]
```
in `Channel.py` I have a class called `ChannelFields` that defines constants for those keys. Maybe we can use it to access fields and keys in  `cln_channel` and potentially do something similar for `lnd`?  but maybe that is also a bit overengineered?

VasanthManiVasi replied:
Thanks! Replaced the hard-coded fields with ChannelFields in a174641
I'm not sure if we should define something similar to ChannelFields for other formats, I think it might be nicer to convert other formats to the cln format (ChannelFields) as it would prevent a lot of overlap which can happen if we define a ChannelFields for each format separately. Could you please share your thoughts about it?

renepickhardt replied:
I think that is reasonable. for now. The fields of other implementations are only being used here. I think on the long term if we also had class called `LNDChannelFields` we could have an automated test to see if lnd changed their API against our class but I think that is overkill. So I think the current fix is ok

renepickhardt reviewed (changes requested):
This looks very good! besides the missing fields I have only a few stylistic suggestions and tiny questions. 

I also very much like how you made it backwards compatible. I was initially thinking of having also a LNDCHannel Base class from which all the others derrive (and I did not like my idea which is why I didn't support lnd yet). but I think the way it is being done here is much more elegant!

Btw does anyone know if there is a checker available that something is the correct lnd or cln json format? If it was we could basically do the check before conversion but I guess that is really unneccesary as the JSON might very well change at some point in time

VasanthManiVasi commented:
Thanks a lot for the review and suggestions!

DescribeGraph.json in LND does not seem to provide channel flags and message flags. Should it be None or set to 0 (since they're bit flags)?

For features, since they're not provided along with the RoutingPolicy of each node in an LND channel edge, we have to search for the corresponding channel node(s) by pubkey in the "nodes" field of DescribeGraph.json (which can be potentially slow as we have to search nodes everytime for each channel direction that we are adding), and then reverse map the dict to feature bits following this: https://github.com/lightningnetwork/lnd/blob/master/lnwire/features.go
Could you please share your thoughts about it?

renepickhardt commented:
> DescribeGraph.json in LND does not seem to provide channel flags and message flags. Should it be None or set to 0 (since they're bit flags)?

I think the python way of doing it would be `None`  though I am pretty sure that the `message flag` are not of importance for us. Maybe we should just remove them from `Channel`? I just wrapped everything that I saw from `cln` without thinking. So I guess the best is to use `None` for now and maybe even deprecate this property.


> For features, since they're not provided along with the RoutingPolicy of each node in an LND channel edge, we have to search for the corresponding channel node(s) by pubkey in the "nodes" field of DescribeGraph.json (which can be potentially slow as we have to search nodes everytime for each channel direction that we are adding), and then reverse map the dict to feature bits following this: https://github.com/lightningnetwork/lnd/blob/master/lnwire/features.go

yeah I think we could just parse the nodes from describe graph and get the features of the nodes and do it the way that you suggested. I think runtime of the import is not a concern (unless we get really slow) as that should hopefully happen only once.

VasanthManiVasi commented:
Yup, I think it would be better to remove channel fields that are not needed for us..
I could be wrong but I think message flags, channel flags and features might not contribute much information for pickhardt payments, so can they be safely removed from ChannelFields?

renepickhardt commented:
let us keep the features. When we upgrade channels to PTLCs we might be in a situation that we can only send the payment via legacy channels or PTLC channels. While it is not clear how we upgrade I would assume a feature bit will be set so Ican forsee why this will be useful in the futre. but we can remove messages I think

VasanthManiVasi commented:
I've removed channel flags and added support for features, but it's very slow as it has to linearly search for nodes (to get their features) each time when adding channels.
It took `162.63` seconds for a channel graph with `13536` nodes and `74496` channels. @renepickhardt

renepickhardt commented on pickhardtpayments/ChannelGraph.py:
```diff
@@ -46,3 +53,80 @@ def get_channel(self, src: str, dest: str, short_channel_id: str):
+            cln_channel[LND_CLN_POLICY_MAP[key]] = val
+
+    def _find_node(pubkey, nodes_list):
+        for node in nodes_list:
```
can't you initially just load all nodes into a `dict` with `pubkey` as index and then just fetch them instead of going linearly through the list everytime you need it?

VasanthManiVasi replied:
Done in 9c733e1

renepickhardt reviewed (changes requested):
I think you can index nodes in a `dict` and avoid linear pass through the list for every time that you want to fetch and element.

VasanthManiVasi commented:
Woops, forgot about that.. Thanks!

renepickhardt commented:
LGTM. Thank you very much! Need to run / test code. Not sure if I will be able to merge this week.

If anyone else wants to do a review that will be great!
